packages/vite/src/node/preview.ts

```diff
@@ -83,18 +89,49 @@ export async function preview(
     'production',
   )
 
+  const options = config.preview
+  const logger = config.logger
+
   const app = connect() as Connect.Server
+
+  const hostname = await resolveHostname(options.host)
+  const port = options.port ?? 4173
+  const protocol = options.https ? 'https' : 'http'
+
   const httpServer = await resolveHttpServer(
     config.preview,
     app,
     await resolveHttpsConfig(config.preview?.https),
   )
   setClientErrorHandler(httpServer, config.logger)
 
+  const serverPort = await httpServerStart(httpServer, {
+    port,
+    strictPort: options.strictPort,
+    host: hostname.host,
+    logger,
+  })
+
+  const resolvedUrls = await resolveServerUrls(
+    httpServer,
+    config.preview,
+    config,
+  )
+
+  const server: PreviewServer = {
+    config,
+    middlewares: app,
+    httpServer,
+    resolvedUrls,
+    printUrls() {
+      printServerUrls(resolvedUrls, options.host, logger.info)
+    },
+  }
+
   // apply server hooks from plugins
   const postHooks: ((() => void) | void)[] = []
   for (const hook of config.getSortedPluginHooks('configurePreviewServer')) {
-    postHooks.push(await hook({ middlewares: app, httpServer }))
+    postHooks.push(await hook(server))
```

Comment:
`configurePreviewServer` hooks now runs after the server started to listen on the port.
I'm not sure if this is safe or not. But I think it's better to align with the dev server.

How about adding `PreviewServerForHook` type that has a nullable `resolvedUrls` property? And then use `PreviewServerForHook` for the hooks and use `PreviewServer` for the return type of `preview` function for compatibility.

Reply:
Ah you're right. I think `PreviewServerForHook` makes sense, but for the next major, maybe we can make it nullable by default so the types are cleaner.

With this change, I think this PR is good too.

Reply:
I think we could merge this in 4.3. @sapphi-red if you agree, you can directly commit your suggestion and merge this one. The alignment is quite uncontroversial to me, we shouldn't need a meeting to review this one IMO.

Reply:
I rebased on main and pushed some commits 👍